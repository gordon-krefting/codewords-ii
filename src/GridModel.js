/* */

export class GridModel {
	
	constructor(gridPattern) {

		this.cells = [];

		for (let rowPattern of gridPattern.trim().split("\n")) {
			let row = [];
			this.cells.push(row);
			for (let cellPattern of rowPattern.trim()) {
				if (cellPattern === '-') {
					row.push(new BlackCell());
				} else if (cellPattern === '.') {
					row.push(new WhiteCell());
				} else if ("A" <= cellPattern && cellPattern <= "Z") {
					row.push(new FixedCell(cellPattern));
				} else {
					throw new Error("Unknown grid symbol:" + cellPattern);
				}
			}
		}
  }

}


export class WhiteCell {
	constructor(label = '', shaded=false) {
		this.label = label;
		this.shaded = shaded;
	}
  toString() {
		return "-";
	}
}

export class BlackCell {
  toString() {
		return "X";
	}
}

export class FixedCell {
	constructor(value, label = '', shaded=false) {
		this.value = value;
		this.label = label;
		this.shaded = shaded;
	}
  toString() {
		return this.value;
	}
}