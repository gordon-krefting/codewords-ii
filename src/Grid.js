import React from 'react';

import {FixedCell, WhiteCell, BlackCell} from './GridModel.js';

export default class Grid extends React.Component {
	render() {
		var o = [];
		let cells = this.props.gridModel.cells;

		for (let i=0; i<cells.length;i++) {
      var p = [];
      for (let j=0; j<cells[i].length;j++) {
      	// replace these divs with a component
      	let cell = cells[i][j];
      	if (cell instanceof BlackCell) {
      		p.push(
	          <div className="square black-square" key={`grd-${i}-${j}`} />
					);
      	} else if (cell instanceof FixedCell) {
      		p.push(
      			<FixedSquare value={cell.value} key={`grd-${i}-${j}`}/>
      		);
      	} else if (cell instanceof WhiteCell) {
      		p.push(
      			<WhiteSquare key={`grd-${i}-${j}`}/>
      		);
      	}
      }
      o.push(<div className="row" key={"grd"+i}>{p}</div>);
    }
		return (<div className="grid">{o}</div>);
	}

}


class FixedSquare extends React.Component {
	render() {
		return (<div className="square fixed-square"><div className="label">{this.props.label}</div><div className="value">{this.props.value}</div></div>);
	}
}

class WhiteSquare extends React.Component {
	render() {
		return (<div className="square white-square"><div className="label">{this.props.label}</div><div className="value"></div></div>);
	}
}

