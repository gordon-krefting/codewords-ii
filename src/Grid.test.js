import {GridModel} from './GridModel.js';
import Grid from './Grid.js';
import React from 'react';
import renderer from 'react-test-renderer';

let gridPattern = `
	-...-
	...B-
	.A.C.
	-....
	-...-
`;
let gridModel = new GridModel(gridPattern);

test('renders correctly', () => {
  const tree = renderer
    .create(<Grid gridModel={gridModel}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});