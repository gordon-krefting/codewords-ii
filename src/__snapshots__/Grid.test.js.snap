// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="grid"
>
  <div
    className="row"
  >
    <div
      className="square black-square"
    />
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square black-square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square fixed-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      >
        B
      </div>
    </div>
    <div
      className="square black-square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square fixed-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      >
        A
      </div>
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square fixed-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      >
        C
      </div>
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="square black-square"
    />
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="square black-square"
    />
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square white-square"
    >
      <div
        className="label"
      />
      <div
        className="value"
      />
    </div>
    <div
      className="square black-square"
    />
  </div>
</div>
`;
